event_sourcery/event_store/in_memory.py

```diff
@@ -0,0 +1,184 @@
+from collections import defaultdict
+from contextlib import contextmanager
+from copy import deepcopy
+from dataclasses import dataclass, field
+from operator import getitem
+from typing import ContextManager, Dict, Generator, Iterator
+
+from typing_extensions import Self
+
+from event_sourcery.event_store import EventStoreFactory
+from event_sourcery.event_store.event import RawEvent
+from event_sourcery.event_store.exceptions import (
+    AnotherStreamWithThisNameButOtherIdExists,
+    ConcurrentStreamWriteError,
+)
+from event_sourcery.event_store.factory import no_filter
+from event_sourcery.event_store.interfaces import (
+    OutboxFiltererStrategy,
+    OutboxStorageStrategy,
+    StorageStrategy,
+)
+from event_sourcery.event_store.stream_id import StreamId
+from event_sourcery.event_store.versioning import NO_VERSIONING, Versioning
+
+
+@dataclass
+class Storage:
+    events: list[RawEvent] = field(default_factory=list, init=False)
+    _data: dict[str, list[RawEvent]] = field(default_factory=dict, init=False)
+    _names: dict[str | None, str] = field(default_factory=dict, init=False)
+    _versions: dict[str | None, dict[str, int | None]] = field(
+        default_factory=lambda: defaultdict(dict), init=False,
+    )
+
+    def _assert_stream_name(self, key: str, name: str | None) -> None:
+        if not name:
+            return
+
+        if name not in self._names:
+            self._names[name] = key
+
+        if self._names[name] != key:
+            raise AnotherStreamWithThisNameButOtherIdExists()
+
+    def _key_from_stream_id(self, stream_id: StreamId) -> str:
```

Comment:
Why won't we use the object of type `StreamId` directly as a key? It seems to be a perfect Value Object 🤔 
For a reason unknown to me I can see `StreamUUID` and `StreamId` being declared with `eq=False` which makes it impossible to use as a key in the first place. Still, why is it defined like this? Otherwise, I believe it should require minimal tuning to make `StreamId` work as a key.

Reply:
Another possible approach to make this class a bit smaller - write a wrapper class over `StreamId` that would take care about comparison and stuff like `AnotherStreamWithThisNameButOtherIdExists()` - WDYT?

Reply:
`eq=False` was for StreamUUID (to use UUID eq). With StreamId you are right, we can make it `eq=True` and simplify storage.

Reply:
done